experimental/oak_baremetal_channel/src/frame.rs

```diff
@@ -0,0 +1,161 @@
+            anyhow::Error::msg(format!(
+                "could convert the frame length usize to {:?}",
+                core::any::type_name::<Length>()
+            ))
+        })
+    }
+}
+
+impl TryFrom<Frame> for Vec<u8> {
+    type Error = anyhow::Error;
+    fn try_from(frame: Frame) -> Result<Self, Self::Error> {
+        let frame_length = frame.len()?;
+        let mut frame_bytes: Vec<u8> = Vec::with_capacity(frame_length.into());
+
+        frame_bytes.extend_from_slice(&[0; PADDING_SIZE]);
+        frame_bytes.extend_from_slice(&frame_length.to_le_bytes());
+        frame_bytes.extend_from_slice(&frame.flags.bits.to_le_bytes());
+        frame_bytes.extend_from_slice(&frame.body);
+
+        Ok(frame_bytes)
+    }
+}
+
+pub struct Framed<T: Read + Write> {
+    inner: T,
+}
+
+impl<T> Framed<T>
+where
+    T: Read<Error = anyhow::Error> + Write<Error = anyhow::Error>,
+{
+    pub fn new(socket: T) -> Self {
+        Self { inner: socket }
+    }
+
+    pub fn read_frame(&mut self) -> anyhow::Result<Frame> {
+        let _padding = {
+            let mut padding_bytes = [0; PADDING_SIZE];
+            self.inner.read_exact(&mut padding_bytes)?;
+            padding_bytes
+        };
+        let length: usize = {
+            let mut length_bytes = [0; LENGTH_SIZE];
+            self.inner.read_exact(&mut length_bytes)?;
+            let length = Length::from_le_bytes(length_bytes).into();
+            if length > MAX_SIZE {
+                return Err(anyhow::Error::msg("frame exceeds the maximum frame size"));
+            };
+            length
+        };
+        let flags = {
+            let mut flags_bytes = [0; FLAGS_SIZE];
+            self.inner.read_exact(&mut flags_bytes)?;
+            Flags::from_bits_truncate(u16::from_le_bytes(flags_bytes))
+        };
+
+        let body = {
+            let body_length: usize = length - BODY_OFFSET;
+            let mut body: Vec<u8> = vec![0; body_length];
+            self.inner.read_exact(&mut body)?;
+            body
+        };
+
+        Ok(Frame { flags, body })
+    }
+
+    pub fn write_frame(&mut self, frame: Frame) -> anyhow::Result<()> {
+        let frame_bytes: Vec<u8> = frame.try_into()?;
+        self.inner.write_all(&frame_bytes)?;
+        self.inner.flush()?;
+        Ok(())
+    }
+}
+
+pub fn bytes_into_frames(data: Vec<u8>) -> Vec<Frame> {
+    let mut frames: Vec<Frame> = data
+        .chunks(MAX_BODY_SIZE)
+        // TODO(#2848): It'd be nice if we didn't have to reallocate here.
+        // We may be able to avoid doing so by making [`Frame`] use
+        // slices and lifetimes. Or alternatively use the Bytes crate for
+        // reference counting.
+        .map(|frame_body| Frame {
+            flags: Flags::default(),
+            body: frame_body.to_vec(),
+        })
+        .collect();
+
+    frames.first_mut().map(|frame| {
+        frame.flags.set(Flags::START, true);
+        Some(())
+    });
```

Comment:
Using map to perform side effects is not idiomatic. In fact, the compiler may just decide to optimize this away since the result is not actually used. Also, first and last are not expected to fail (i.e. they should never be None), so it seems fine to just unwrap / expect, instead of silently ignoring potential errors. Well, apart from empty data as input, but I would say that's in fact an error too (and maybe it could even be handled separately).

Reply:
> Using map to perform side effects is not idiomatic
haha I would be inclined to agree, I too feel that map should not be used for side effects. (though imo it's unlikely compilers would actually make that assumption)

Anyways, one part that does not agree with this sentiment is apparently "cargo clippy". Originally this used to be a "and_then", but it clippy flagged that as "error: using `Option.and_then(|x| Some(y))`, which is more succinctly expressed as `map(|x| y)`".

> Also, first and last are not expected to fail (i.e. they should never be None), so it seems fine to just unwrap / expect, instead of silently ignoring potential errors. Well, apart from empty data as input, but I would say that's in fact an error too (and maybe it could even be handled separately).

Yh, empty data would be the case where unwrap would lead to panic. I guess we can handle that separately, but originally decided not to, as the unwrap/expect here would be reliant behaviors a few lines up.

Reply:
`and_then` should also not be used for side effects only. That would be done by an `if let` or a `match`.